node/node.go

```diff
@@ -1027,48 +1026,54 @@ func (n *nodeImpl) NodeInfo() types.NodeInfo {
 }
 
 // startStateSync starts an asynchronous state sync process, then switches to fast sync mode.
-func startStateSync(ssR *statesync.Reactor, bcR cs.FastSyncReactor, conR *cs.Reactor,
-	stateProvider statesync.StateProvider, config *cfg.StateSyncConfig, fastSync bool,
-	stateStore sm.Store, blockStore *store.BlockStore, state sm.State) error {
-	ssR.Logger.Info("starting state sync...")
-
-	if stateProvider == nil {
-		var err error
-		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
-		defer cancel()
-		stateProvider, err = statesync.NewLightClientStateProvider(
-			ctx,
-			state.ChainID, state.Version, state.InitialHeight,
-			config.RPCServers, light.TrustOptions{
-				Period: config.TrustPeriod,
-				Height: config.TrustHeight,
-				Hash:   config.TrustHashBytes(),
-			}, ssR.Logger.With("module", "light"))
-		if err != nil {
-			return fmt.Errorf("failed to set up light client state provider: %w", err)
-		}
+func startStateSync(
+	ssR statesync.SyncReactor,
+	bcR cs.FastSyncReactor,
+	conR cs.ConsSyncReactor,
+	config *cfg.StateSyncConfig,
+	fastSync bool,
+	state sm.State,
+	eb *types.EventBus) error {
+	ssR.GetLogger().Info("starting state sync...")
+
+	if err := ssR.InitStateProvider(state, config.RPCServers, light.TrustOptions{
+		Period: config.TrustPeriod,
+		Height: config.TrustHeight,
+		Hash:   config.TrustHashBytes()}, ssR.GetLogger().With("module", "light")); err != nil {
+		return fmt.Errorf("failed to set up light client state provider: %w", err)
+	}
+
+	// at the beginning of the statesync start, we use the initialHeight as the event height
+	// because of the statesync doesn't have the concreate state height before fetched the snapshot.
+	d := types.EventDataStateSyncStatus{Complete: false, Height: state.InitialHeight}
+	if err := eb.PublishEventStateSyncStatus(d); err != nil {
+		ssR.GetLogger().Error("failed to emit the statesync start event", "err", err)
 	}
 
 	go func() {
-		state, err := ssR.Sync(context.TODO(), stateProvider, config.DiscoveryTime)
+		state, err := ssR.Sync(context.TODO(), config.DiscoveryTime)
 		if err != nil {
-			ssR.Logger.Error("state sync failed", "err", err)
+			ssR.GetLogger().Error("state sync failed", "err", err)
```

Comment:
codesmell. Don't log here. The reactor should log on error.

Reply:
but it is a different goroutine, the caller might not be able to receive the return error.